sdk/storage/azure-storage-blob/azure/storage/blob/_shared/policies.py

```diff
@@ -0,0 +1,594 @@
+            except:  # pylint: disable=bare-except
+                pass
+            for chunk in iter(lambda: data.read(4096), b""):
+                md5.update(chunk)
+            try:
+                data.seek(pos, SEEK_SET)
+            except (AttributeError, IOError):
+                raise ValueError("Data should be bytes or a seekable file-like object.")
+        else:
+            raise ValueError("Data should be bytes or a seekable file-like object.")
+
+        return md5.digest()
+
+    def on_request(self, request, **kwargs):
+        # type: (PipelineRequest, Any) -> None
+        validate_content = request.context.options.pop('validate_content', False)
+        if validate_content and request.http_request.method != 'GET':
+            computed_md5 = encode_base64(StorageContentValidation.get_content_md5(request.http_request.data))
+            request.http_request.headers[self.header_name] = computed_md5
+            request.context['validate_content_md5'] = computed_md5
+        request.context['validate_content'] = validate_content
+
+    def on_response(self, request, response, **kwargs):
+        if response.context.get('validate_content', False) and response.http_response.headers.get('content-md5'):
+            computed_md5 = request.context.get('validate_content_md5') or \
+                encode_base64(StorageContentValidation.get_content_md5(response.http_response.body()))
+            if response.http_response.headers['content-md5'] != computed_md5:
+                raise AzureError(
+                    'MD5 mismatch. Expected value is \'{0}\', computed value is \'{1}\'.'.format(
+                        response.http_response.headers['content-md5'], computed_md5),
+                    response=response.http_response
+                )
+
+
+class StorageRetryPolicy(HTTPPolicy):
+    """
+    The base class for Exponential and Linear retries containing shared code.
+    """
+
+    def __init__(self, **kwargs):
+        self.total_retries = kwargs.pop('retry_total', 10)
+        self.connect_retries = kwargs.pop('retry_connect', 3)
+        self.read_retries = kwargs.pop('retry_read', 3)
+        self.status_retries = kwargs.pop('retry_status', 3)
+        self.retry_to_secondary = kwargs.pop('retry_to_secondary', False)
+        super(StorageRetryPolicy, self).__init__()
+
+    def _set_next_host_location(self, settings, request):  # pylint: disable=no-self-use
+        """
+        A function which sets the next host location on the request, if applicable.
+
+        :param ~azure.storage.models.RetryContext context:
+            The retry context containing the previous host location and the request
+            to evaluate and possibly modify.
+        """
+        if settings['hosts'] and all(settings['hosts'].values()):
+            url = urlparse(request.url)
+            # If there's more than one possible location, retry to the alternative
+            if settings['mode'] == LocationMode.PRIMARY:
```

Comment:
shouldn't this method be defensive and check whether retry to secondary is desired?

Reply:
This method is already only run if the policy is configured to retry to secondary (off by default)
- See line 379 for the settings configuration
- See line 442 for the settings check before changing the hostname